### ISSUE TITLE
json middleware does not work on content types with a `+`

### ISSUE THREAD

**Issue by jeswr** [expressjs/body-parser#519]
The json middleware does not parse results as json if the content type is anything but `application/json`, I would expect it to also work on content types like `application/ld+json`. For instance:

```ts
import express from 'express';
const app = express();
app.use(express.json());

app.post('/', async (req, res) => {
    console.log("Data posted", await req.body);
});

app.listen(3000, () => {
    console.log('Listening on port 3000');
    fetch('http://localhost:3000', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            message: 'Hello world',
        }),
    });
});
```

returns

```
Listening on port 3000
Data posted { message: 'Hello world' }
```

On the other hand

```ts
import express from 'express';
const app = express();
app.use(express.json());

app.post('/', async (req, res) => {
    console.log("Data posted", await req.body);
});

app.listen(3000, () => {
    console.log('Listening on port 3000');
    fetch('http://localhost:3000', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/ld+json',
        },
        body: JSON.stringify({
            message: 'Hello world',
        }),
    });
});
```

returns

```
Listening on port 3000
Data posted {}
```

**Comment by wesleytodd** [comment]
Moved this to the correct repo.

How would you expect to see the `req.body` with `ls+json`? I would expect an array where each line is an item parsed as json individually. I feel personally like this would be a **new** middleware exported by this package if we chose to add support? Not sure I have ever seen anyone use ldjson as input to an http api before, so maybe would be good to give a description of your use case as well?

**Comment by jonchurch** [comment]
Folks can do this today with 
```js
app.use(express.json({type: ['application/*+json', 'application/json']}))
```

Unfortunately `"application/*json` doesn't work. That is a [`type-is`](https://github.com/jshttp/type-is/tree) limitation. I agree that we should have a way to allow parsing of all json subtypes.

If you receive input with `+json` media type it is understood to be safe to parse as a JSON doc because it must be one. NDJSON cannot be parsed as a single JSON doc so it has a distinct mimetype `application/x-ndjson`, for example.

I am open to this being the default behavior. Im sure 99% of people are just using application/json, so it's a small gain which wouldn't be worth it for correctness alone if it had some downside. Im not aware of any real ones though except additional load/risk parsing payloads you weren't expecting. 

They should still be JSON though and you'd have to go out of your way to enable parsing a valid json doc if you did want to currently.

**Comment by wesleytodd** [comment]
> I am open to this being the default behavior

I don't think this should be a default behavior for `.json`. I don't think api's accepting ld-json are common (elastic bulk updates is the only one I can think of) and I think it is pretty acceptable to have to turn it on. If a client started sending it and my server started crashing because I expected always an object but `.json` started populating `.body` with an array I would be pretty miffed.

**Comment by jonchurch** [comment]
Hmmm ldjson isnt necessarily an array. But an array is still valid JSON so thats an API implementation detail you can run into regardless of your specific json sub mimetype. 

Lets make sure we are talking about the same thing. application/ld+json is [Linked Data JSON](https://www.w3.org/TR/2014/REC-json-ld-20140116/#interpreting-json-as-json-ld), its still a valid JSON blob, or potentially a top level array of ldjson objects. 

But the toplevel array possibility is universal to all valid JSON so far as I know, that possibility doesnt relate to ldjson. 

I agree it could be confusing, but I am still looking for an actual downside. It is equally confusing that we dont parse valid JSON objects

**Comment by wesleytodd** [comment]
> Lets make sure we are talking about the same thing. 

I think I just accidentally wrote ld and not nd. Anyway, can't keep up with this convo right now I just wanted to drop a note that doing this by default is something we should avoid if we can imo.

**Comment by jonchurch** [comment]
Heard. Ndjson has its own mimetype and would not be impacted by the suggested change. 

`application/x-ndjson`

Mime has the + subclasses to deal with OP's topic specifically. 

Any `+json` should be fully parseable by `JSON.parse`

`application/x-ndjson` would not be matched by something like `*+json` in the proposed changes.

**Comment by jeswr** [comment]
+1 on @jonchurch's points

> How would you expect to see the req.body with ls+json? I feel personally like this would be a new middleware exported by this package if we chose to add support?

Just the result of calling `JSON.parse` on the body, no different to the current behavior of the handler on the `application/json` content type; since "Any `+json` should be fully parseable by `JSON.parse`"

> Not sure I have ever seen anyone use ldjson as input to an http api before, so maybe would be good to give a description of your use case as well?

I'm currently working on a project which involves conversational Web agents that are posting RDF to one another, granted I'm parsing them to a [dataset](https://rdf.js.org/dataset-spec/#datasetcore-interface) most of the time.

This would come up in production settings when implementing specs like [Verifiable Credentials](https://www.w3.org/TR/vc-data-model-2.0/) where there was a division between spec writers on whether to use plain JSON or RDF representations for the VCs. The compromise was that framed JSON-LD would be used. The specification uses `POST`ing of framed JSON-LD in order to [issue credentials](https://w3c-ccg.github.io/vc-api/#issue-credential).
